### PR TITLE
added pagination

### PR DESCRIPTION
**Pull request by violet360**
route  -> /listForwards/:offset/:maxLen/:reverse
returns a subarray from list of forwards
offset: index of the subarray in the forwardingList
maxLen: ending index of the subarray = (offset + maxLen) - 1
reverse: if true will return reversed subarray

### CODE REVIEW

**Comment by violet360** [comment]
whenever an element in subarray goes out of range, it gets omitted from the return array

**Review by ShahanaFarooqui — CHANGES_REQUESTED** [review]
Thanks for the PR. However, we want to explore the option to cache the data after the first request. Please advice if we should always fetch the data from server or cache for consecutive requests.

**Comment by violet360** [comment]
great idea actually!
I was reluctant at first but I think caching should be done.
the prime problem of what I think with caching is old data, now I don't think we need real time data in list forwards, some level of staleness is fine, so if RTL requests the listForwards from cltRest, within 10 minutes of the previous request then we will fetch from cache, else we will fetch from the clt, and update the cache.

**Comment by ShahanaFarooqui** [comment]
Thanks for the advice and great observation. I would suggest:

- To keep caching time up to 5 minutes not 10.
- And to fetch the fresh data from CLT whenever the user requests for the first page of the list, even when the data is not 5 minutes old in the cache. It will ensure that the user gets latest forwards details also.

Now, that we have agreed on caching, please advice on how can we implement it, ie the caching mechanism, data details etc.

**Comment by saubyk** [comment]
@violet360 @ShahanaFarooqui 
While implementing caching is a good idea, I think we can keep it for enhancement at a later date. I would like to proceed with the current solution, right now.

since C-Lightning-REST executes at the same node as C-Lightning, even if it has to make repeated calls, it may not create a significant processing overhead, right now.

File: controllers/channel.js [hunk 1]
```diff
@@ -413,11 +413,32 @@ exports.closeChannel = (req,res) => {
 exports.listForwards = (req,res) => {
     function connFailed(err) { throw err }
     ln.on('error', connFailed);
-
+    var {offset, maxLen, reverse} = req.params
```
Comment (saubyk) [inline controllers/channel.js:416]: this should be:
`var {offset, maxLen, reverse} = req.query`

**Review by saubyk — CHANGES_REQUESTED** [review]

File: controllers/channel.js (hunk 1 above)
Comment (violet360) [reply controllers/channel.js:416]: used req.params, cause I was parsing it via params in URL, wanted to keep it homogeneous with rest of the code 😅  , you sure you want me to remove the optional params in routes and change it to query ?

File: routes/channel.js [hunk 2]
```diff
@@ -19,6 +19,6 @@ router.delete('/closeChannel/:id', tasteMacaroon, channelController.closeChannel
 router.get('/localRemoteBal', tasteMacaroon, localRemoteBalController.localRemoteBal);
 
 //Get the list of htlcs forwarded
-router.get('/listForwards', tasteMacaroon, channelController.listForwards);
+router.get('/listForwards/:offset?/:maxLen?/:reverse?', tasteMacaroon, channelController.listForwards);
```
Comment (violet360) [inline routes/channel.js:22]: the optional params I need to remove

File: controllers/channel.js (hunk 1 above)
Comment (saubyk) [reply controllers/channel.js:416]: The standard which we are following (for this library) is to pass all the optional parameters for GET calls as query params.

File: routes/channel.js (hunk 2 above)
Comment (saubyk) [reply routes/channel.js:22]: yes

**Pull request closed without merging**